README.md
# Cover

> A Java library that mimics CSS `object-fit` image scaling behavior

[![Maven Central](https://img.shields.io/maven-central/v/io.taig/object-fit)](https://search.maven.org/artifact/io.taig/object-fit)

A Java AWT based image scaling library that is inspired by <a href="https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit?retiredLocale=de">CSS `object-fit`</a> scaling modes. This library allows to scale images in `cover`, `contain` and `fill` modes, where `cover` mode is the noteworthy feature.

Additionally, this library automatically handles image rotation from JPEG Exif data if available.

_object-fit_ is well suited for thumbnail generation but works just as well with larger images.

## Installation

### sbt

```sbt
libraryDependencies += "io.taig" % "object-fit" % "x.y.z"
```

### Maven

```xml
<dependency>
  <groupId>io.taig</groupId>
  <artifactId>object-fit</artifactId>
  <version>x.y.z</version>
</dependency>
```

### Gradle

```groovy
implementation("io.taig:object-fit:x.y.z")
```

## Usage

```java
import io.taig.object.fit.ObjectFit;
import java.io.File;

InputStream input = ...

final ObjectFit.Options = ObjectFit.options()
  .mode(ObjectFit.Mode.COVER)
  .size(500)
  .format("jpg");

ObjectFit.toBufferedImage(input, options);
ObjectFit.write(input, options, ...);
```

## Samples

### Cover

```java
ObjectFit.options()
  .mode(ObjectFit.Mode.COVER)
  .size(250, 150)
  .format("webp")
```

<table>
  <tr>
    <th>Input</th>
    <th>Output</th>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.1.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.1.1.result.webp" /></td>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.2.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.1.2.result.webp" /></td>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.3.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.1.3.result.webp" /></td>
  </tr>
</table>

<hr />

```java
ObjectFit.options()
  .mode(ObjectFit.Mode.COVER)
  .size(250, 150)
  .format("webp")
  .scaleUp()
  ```

<table>
  <tr>
    <th>Input</th>
    <th>Output</th>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.3.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.2.3.result.webp" /></td>
  </tr>
</table>

### Fill

```java
ObjectFit.options()
  .mode(ObjectFit.Mode.FILL)
  .size(250, 250)
  .format("png")
```

<table>
  <tr>
    <th>Input</th>
    <th>Output</th>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.1.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.3.1.result.png" /></td>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.2.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.3.2.result.png" /></td>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.3.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.3.3.result.png" /></td>
  </tr>
</table>

<hr />

```java
ObjectFit.options()
  .mode(ObjectFit.Mode.FILL)
  .size(250, 250)
  .format("png")
  .scaleUp()
```

<table>
  <tr>
    <th>Input</th>
    <th>Output</th>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.3.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.4.3.result.png" /></td>
  </tr>
</table>

### Contain

```java
ObjectFit.options()
  .mode(ObjectFit.Mode.CONTAIN)
  .size(250, 250)
  .format("jpg")
```

<table>
  <tr>
    <th>Input</th>
    <th>Output</th>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.1.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.5.1.result.jpg" /></td>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.2.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.5.2.result.jpg" /></td>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.3.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.5.3.result.jpg" /></td>
  </tr>
</table>

<hr />

```java
ObjectFit.options()
  .mode(ObjectFit.Mode.CONTAIN)
  .size(400, 500)
  .format("jpg")
  .scaleUp()
```

<table>
  <tr>
    <th>Input</th>
    <th>Output</th>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.1.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.6.1.result.jpg" /></td>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.2.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.6.2.result.jpg" /></td>
  </tr>
  <tr>
    <td><img src="/modules/samples/src/main/resources/otter.3.jpg" /></td>
    <td><img src="/modules/samples/src/main/resources/otter.6.3.result.jpg" /></td>
  </tr>
</table>

CHANGELOG.md
# Changelog

## 0.1.3

_2022-12-05_

* Improve MetadataInputStream performance

## 0.1.2

_2022-12-04_

* Make ObjectFit.Options public

## 0.1.1

_2022-12-04_

* Fix ObjectFit.Mode visibility

## 0.1.0

_2022-12-04_

* Rename to `object-fit` and add more scaling modes

## 0.0.3

_2022-12-03_

* Only extract metadata and perform image rotation for jpgs

## 0.0.2

_2022-12-03_

* Improve metadata extraction (fixes webp decoding issue)
* Add image encoding type parameter to `fit` method

## 0.0.1

_2022-11-28_

* Initial release

modules/samples/src/main/scala/io/taig/objectfit/Samples.scala
package io.taig.objectfit

import cats.effect.{IO, IOApp}
import io.taig.objectfit.ObjectFit

import java.io.File

object Samples extends IOApp.Simple:
  override def run: IO[Unit] =
    IO(System.setProperty("java.awt.headless", "true")) *>
      one("/otter.1.jpg", "otter.1.1.result.webp") *>
      one("/otter.2.jpg", "otter.1.2.result.webp") *>
      one("/otter.3.jpg", "otter.1.3.result.webp") *>
      two("/otter.3.jpg", "otter.2.3.result.webp") *>
      three("/otter.1.jpg", "otter.3.1.result.png") *>
      three("/otter.2.jpg", "otter.3.2.result.png") *>
      three("/otter.3.jpg", "otter.3.3.result.png") *>
      four("/otter.3.jpg", "otter.4.3.result.png") *>
      five("/otter.1.jpg", "otter.5.1.result.jpg") *>
      five("/otter.2.jpg", "otter.5.2.result.jpg") *>
      five("/otter.3.jpg", "otter.5.3.result.jpg") *>
      six("/otter.1.jpg", "otter.6.1.result.jpg") *>
      six("/otter.2.jpg", "otter.6.2.result.jpg") *>
      six("/otter.3.jpg", "otter.6.3.result.jpg")

  def target(name: String): File = new File(s"./modules/samples/src/main/resources/$name")

  def one(source: String, destination: String): IO[Unit] = IO.blocking {
    val options = ObjectFit
      .options()
      .mode(ObjectFit.Mode.COVER)
      .size(250, 150)
      .format("webp")

    ObjectFit.write(getClass.getResourceAsStream(source), options, target(destination))
  }

  def two(source: String, destination: String): IO[Unit] = IO.blocking {
    val options = ObjectFit
      .options()
      .mode(ObjectFit.Mode.COVER)
      .size(250, 150)
      .format("webp")
      .scaleUp()

    ObjectFit.write(getClass.getResourceAsStream(source), options, target(destination))
  }

  def three(source: String, destination: String): IO[Unit] = IO.blocking {
    val options = ObjectFit
      .options()
      .mode(ObjectFit.Mode.FILL)
      .size(250, 250)
      .format("png")

    ObjectFit.write(getClass.getResourceAsStream(source), options, target(destination))
  }

  def four(source: String, destination: String): IO[Unit] = IO.blocking {
    val options = ObjectFit
      .options()
      .mode(ObjectFit.Mode.FILL)
      .size(250, 250)
      .format("png")
      .scaleUp()

    ObjectFit.write(getClass.getResourceAsStream(source), options, target(destination))
  }

  def five(source: String, destination: String): IO[Unit] = IO.blocking {
    val options = ObjectFit
      .options()
      .mode(ObjectFit.Mode.CONTAIN)
      .size(250, 250)
      .format("jpg")

    ObjectFit.write(getClass.getResourceAsStream(source), options, target(destination))
  }

  def six(source: String, destination: String): IO[Unit] = IO.blocking {
    val options = ObjectFit
      .options()
      .mode(ObjectFit.Mode.CONTAIN)
      .size(400, 500)
      .format("jpg")
      .scaleUp()

    ObjectFit.write(getClass.getResourceAsStream(source), options, target(destination))
  }
